crates/typst-cli/src/query.rs

```diff
@@ -0,0 +1,133 @@
+use typst::World;
+use typst_library::meta::ProvideElem;
+use typst_library::prelude::*;
+
+use crate::args::QueryCommand;
+use crate::compile::print_diagnostics;
+use crate::set_failed;
+use crate::world::SystemWorld;
+
+#[derive(Serialize)]
+pub struct SelectedElement {
+    #[serde(rename = "type")]
+    typename: String,
+    attributes: HashMap<EcoString, Value>,
+}
+
+/// Execute a query command.
+pub fn query(command: QueryCommand) -> StrResult<()> {
+    let mut world = SystemWorld::new(&command.common)?;
+    tracing::info!("Starting querying");
+
+    let start = std::time::Instant::now();
+    // Reset everything and ensure that the main file is still present.
+    world.reset();
+    world.source(world.main()).map_err(|err| err.to_string())?;
+
+    let mut tracer = Tracer::default();
+    let result = typst::compile(&world, &mut tracer);
+    let duration = start.elapsed();
+    let warnings = tracer.warnings();
+
+    match result {
+        // Print metadata
+        Ok(document) => {
+            let introspector = Introspector::new(&document.pages);
+
+            if let Some(key) = &command.key {
+                let provided_metadata = introspector
+                    .query(&keyvalue_selector(key))
+                    .iter()
+                    .filter_map(|c| c.field("value"))
+                    .collect::<Vec<_>>();
+                format(&provided_metadata, &command)?;
+            }
+
+            if let Some(selector) = &command.selector {
+                let selected_metadata = introspector
+                    .query(&generic_selector(selector, &world)?)
+                    .into_iter()
+                    .map(|x| SelectedElement {
+                        typename: x.func().name().into(),
+                        attributes: x
+                            .clone()
+                            .into_inner()
+                            .fields()
+                            .map(|(k, v)| (k.clone(), v))
+                            .collect(),
+                    })
+                    .collect::<Vec<_>>();
+                format(&selected_metadata, &command)?;
+            }
+
+            tracing::info!("Processing succeeded in {duration:?}");
+
+            print_diagnostics(&world, &[], &warnings, command.common.diagnostic_format)
+                .map_err(|_| "failed to print diagnostics")?;
+        }
+
+        // Print diagnostics.
+        Err(errors) => {
+            set_failed();
+            tracing::info!("Processing failed");
+
+            print_diagnostics(
+                &world,
+                &errors,
+                &warnings,
+                command.common.diagnostic_format,
+            )
+            .map_err(|_| "failed to print diagnostics")?;
+        }
+    }
+
+    Ok(())
+}
+
+fn generic_selector(description: &str, world: &dyn World) -> StrResult<Selector> {
+    let evaluated_selector = eval_string(
+        world.track(),
+        &format!("selector({description})"),
```

Comment:
there is no need to wrap the string with "selector(..)". instead you can `.cast()` the resulting value to a `Selector` and will coerce things like functions and labels automatically

Reply:
this is handy